Make csi drivers and in-tree drivers share e2e tests

mkimuram opened the pull request:
**What this PR does / why we need it**:
Make csi drivers and in-tree drivers share e2e tests to make test coverage of the csi drivers the same level to that of in-tree drivers.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #68024

**Special notes for your reviewer**:
/sig storage
/sig testing

**Release note**:
```release-note
NONE
```

mkimuram commented:
@msau42 @jsafrane @wongma7 @davidz627 

This PR is to apply https://github.com/kubernetes/kubernetes/pull/66577 to csi drivers.
I will rebase this commit once in-tree one is merged, then the diff will be much smaller.
(I can wait for your review after I rebase.)

dims commented on hack/.golint_failures:
```diff
@@ -751,6 +751,8 @@ test/e2e/scalability
 test/e2e/scheduling
 test/e2e/servicecatalog
 test/e2e/storage
+test/e2e/storage/drivers
```
Do you mind fixing up the lint errors, so we don't increase the technical debt? please.

mkimuram replied:
Thank you for your comment.
Actually, it is added to avoid "should not use dot imports" errors caused by dot importing ginkgo and gomega. I will fix up lint errors other than that.

jsafrane commented:
/ok-to-test

dims commented:
/unassign

davidz627 commented:
/assign
/assign @msau42

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -0,0 +1,339 @@
+
+func testProvisioning(input *provisioningTestInput) {
+	It("should provision storage", func() {
+		TestDynamicProvisioning(input.testCase, input.cs, input.pvc, input.sc)
```
Is there another test suite this could be a part of? I'm not sure it's unique enough to warrant moving it out into its own suite.

Also if we're moving this to the new test suite framework can we make the in-tree drivers consume the tests suite in the new way as well?

mkimuram replied:
@davidz627

Thank you for your comment.

> Is there another test suite this could be a part of? I'm not sure it's unique enough to warrant moving it out into its own suite (but then again maybe it is, I didn't look that carefully).

```TestDynamicProvisioning``` is still used by below tests:
  - test/e2e/storage/regional_pd.go
  - test/e2e/storage/volume_provisioning.go

I expect that all of them will eventually move to this framework to ensure that all driver are tested in the same way.
However, I would like to limit the scope of this PR to make the testsuites of csi_volumes.go and in_tree_volumes.go the same. So, only provisioning tests that is called from csi_volumes.go was ported to this framework.
(There are so many cloud provider dependent tests in these files, so doing it at once will require too much work that is not the original purpose of this PR.)
 
Until then, ```TestDynamicProvisioning``` needs to be called from storage package. So, it is moved to testsuites package to be used in this framework and changed to global scope.


> Also if we're moving this to the new test suite framework can we make the in-tree drivers consume the tests suite in the new way as well?

It is already done in this commit and only adding below one line to test/e2e/storage/in_tree_volumes.go will make in-tree drivers run this provisioning test suite.

https://github.com/kubernetes/kubernetes/pull/68025/files/90dbd631243c70f7db3698410a67ea80b18cbf93#diff-e6441be49bc83f3d2436271f90c14742R53

mkimuram commented:
@msau42 @jsafrane 

I would like to prioritize https://github.com/kubernetes/kubernetes/pull/65829 over this PR.
I noticed that both of them touch the same part of test/e2e/storage/volume_provisioning.go in a different way, which will make merge conflict again. 

This PR would be helpful to make CSI Block volume feature Beta, but it will depend on Block feature to be Beta, which https://github.com/kubernetes/kubernetes/pull/65829 is doing. So, hopefully it will be merged soon.

dims commented:
/uncc

mkimuram commented:
@msau42

Thank you for your reminder. I'm working on it. 
It will take a little bit more time to apply the same refactoring to ```SkipAttach``` test, which was added to csi_volumes.go.

mkimuram commented:
@msau42 

I re-based the commit. PTAL

It seems that tests for csi drivers are now taged with ```[Serial]```, so I implemented the same way as a starting point. I hope that we will be able to make it parallel again, in this commit or in a separate commit, depending on the difficulty.  (I guess that not deleting CRD and making provisioner name and volume resource different between tests would be enough.)

@jsafrane 

To make in-tree drivers and csi drivers share the same framework, I modified and moved  ```SkipAttach``` test to ```testsuites``` package. Could you review it?
I'm not familiar with this feature, but I hope that I understand the intention of this test correctly. (As a result of test in my local environment with ```CSIDriverRegistry=true```, ```SkipAttach``` test for ```csi-hostpath``` looks working fine.)
Also, I would like to confirm if this test should be run against in-tree drivers as well. (We might be able to make it out of the scope of this commit, though.)

mkimuram commented:
@msau42 @jsafrane 

I re-based the commit again. PTAL

mkimuram commented:
@msau42 @jsafrane

I re-based the commit again. PTAL

davidz627 commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -0,0 +1,243 @@
+	provisioner := g.driverInfo.Name
+	suffix := fmt.Sprintf("%s-sc", g.driverInfo.Name)
+
+	// pods should be scheduled on the node
```
not sure what this is for but I don't think its necessary for GCE PD. We can schedule test pods on any node

msau42 replied:
@davidz627 can you help test this PR on the gce environment?

I agree we should not be setting nodes here

davidz627 commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -0,0 +1,243 @@
+
+	nodeZone, ok := node.GetLabels()[kubeletapis.LabelZoneFailureDomain]
+	Expect(ok).To(BeTrue(), "Could not get label %v from node %v", kubeletapis.LabelZoneFailureDomain, node.GetName())
+	parameters := map[string]string{"type": "pd-standard", "zone": nodeZone}
```
we removed the "zone" parameter from the CSI Driver, just leave it unspecified here.

msau42 replied:
Topology support in CSI is still alpha.  We may need to skip the test in multi-zone environments, and add an explicit [Feature] tag for csi topology tests to run in alpha clusters.  cc @verult @ddebroy

davidz627 commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -0,0 +1,243 @@
+	g.controllerServiceAccount = csiServiceAccount(cs, config, "gce-controller", false /* setup */)
+	g.nodeServiceAccount = csiServiceAccount(cs, config, "gce-node", false /* setup */)
+	csiClusterRoleBindings(cs, config, false /* setup */, g.controllerServiceAccount, g.controllerClusterRoles)
+	csiClusterRoleBindings(cs, config, false /* setup */, g.nodeServiceAccount, g.nodeClusterRoles)
```
we're missing the podsecuritypolicy bindings here, they are necessary as of 1.12+ as PSP is enabled by default  on CI

davidz627 commented on test/e2e/storage/testsuites/volume_attachment.go:
```diff
@@ -0,0 +1,232 @@
+	canSkipAttach bool
+}
+
+func testVolumeAttachment(input *volumeAttachmentTestInput) {
```
@verult could you review this file

davidz627 replied:
or @jsafrane, I'm not sure who knows most about these tests

davidz627 reviewed (commented):
Initial pass, will do a more in depth review later.

Were you able to get the tests for both hostpath and GCE PD up and running by any chance?

mkimuram commented:
/retest

mkimuram commented:
@davidz627 

Latest commit should reflect your comments. PTAL

msau42 commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -0,0 +1,243 @@
+	framework.SkipUnlessProviderIs("gce", "gke")
+	// TODO(#62561): Use credentials through external pod identity when that goes GA instead of downloading keys.
+	createGCESecrets(cs, config)
+	framework.SkipUnlessSecretExistsAfterWait(cs, "cloud-sa", config.Namespace, 3*time.Minute)
```
@davidz627 should this be a failure instead of skip?

msau42 commented on test/e2e/storage/testsuites/volume_attachment.go:
```diff
@@ -0,0 +1,232 @@
+}
+
+func testVolumeAttachment(input *volumeAttachmentTestInput) {
+	It("should create VolumeAttachment, when driver doesn't exist", func() {
```
will this test case still work?  Doesn't the test framework already launch the csi driver before this?

msau42 commented:
I would wait one more week on this. @davidz627 and @jsafrane are making many updates to CSI tests right now

mkimuram commented:
@msau42 

> I would wait one more week on this. @davidz627 and @jsafrane are making many updates to CSI tests right now

Thank you for letting me know it. I will also wait until it become stable.
Although, I still hope that this refactoring is merged in earlier phase so that we can update driver codes and feature-test codes separately without affecting each other too much.

davidz627 commented:
Sorry about the churn @mkimuram there are some last minute GCE PD CSI test changes that I need to get in to get everything up and running on the driver side. They are not major but they will be moving some things around.

msau42 commented on test/e2e/storage/testsuites/volume_attachment.go:
```diff
@@ -0,0 +1,232 @@
+	tsInfo TestSuiteInfo
+}
+
+var _ TestSuite = &volumeAttachmentTestSuite{}
```
I think we should not add these tests to the general suite.  These tests are more like integration tests, it's checking for implementation details that users do not need to care about.  @jsafrane wdyt?

jsafrane replied:
I want to make sure that A/D controller and kubelet work together to bring up a volume with various CSIDrivers. IMO, this is not possible in integration tests.

mkimuram replied:
@jsafrane 

I create a separate PR for volumeAttachment test. PTAL
https://github.com/kubernetes/kubernetes/pull/69469

It is only enabled for in-tree driver and will be just skipped until volumeAttachment is implemented for in-tree driver.

msau42 replied:
Do we need to explicitly test the creation of VolumeAttachment objects for every CSI Plugin though?  Would it be sufficient to just test that you can launch a pod and read/write to it against any CSI driver?  The CSI driver deployment for non-attachable plugins can leave out the external-attacher sidecar to test this.

jsafrane replied:
> Do we need to explicitly test the creation of VolumeAttachment objects for every CSI Plugin though?

No, it's enough for hostpath CSI driver. The tests creates various CSIDrivers for it. It does not make sense to test it with GCE driver.

pohly commented:
I have preliminary code which deploys the entire set of RBAC rules, daemonset, statefulset, storageclass for a CSI driver from .yaml files. The entities get renamed such that they are unique, so even tests that use the same provisioner can run in parallel.

Let me see whether I can polish that code and push it as a PR tomorrow. It should simplify writing the E2E tests.

msau42 commented:
@mkimuram thanks so much for working on this!

I would leave the volume attachment tests as they are and not move them to the framework. They are not designed to be run against every plugin.

pohly commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -0,0 +1,238 @@
+
+const (
+	csiExternalProvisionerClusterRoleName string = "system:csi-external-provisioner"
+	csiExternalAttacherClusterRoleName    string = "system:csi-external-attacher"
```
These roles are deprecated. Don't use them. Instead each driver deployment must create its own roles.

PR #69868 makes that change. Let's decide about that first, then figure out how to proceed with this PR.

mkimuram replied:
@pohly 

Thank you for your comment.
This PR is focusing on sharing the test with in-tree drivers and csi drivers, not focusing on how csi driver will prepare test, like creating sa, clusterrole, ds, and so on. 

It seems an improvement for specific to how to prepare test for csi-hostpath driver. So, I think that we can improve it separately with this PR.

mkimuram replied:
@pohly 

This PR also aim to split driver logics and test logics, which has already been done in in-tree driver.
RBAC related change would be only in driver logics, so once this PR is merged, any changes for test logics will be able to merge without conflict to driver logic changes including RBAC related changes. (So, I prioritize this PR over my csi rbd driver PR.)

I don't mean to say that this PR should be prioritized over RBAC PR, and I will rebase this PR if RBAC PR is merged before this PR.

pohly commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -0,0 +1,238 @@
+*/
+
+/*
+ * This file defines various csi volume test drivers for TestSuites.
```
How can out-of-tree CSI drivers define their own instance of these shared tests?

They can't extend `drivers/csi.go`, because it is read-only for out-of-tree tests.

IMHO the in-tree CSI tests should be defined the same way. That way we know that the mechanism (whatever it is) works.

mkimuram replied:
This topic should be covered in https://github.com/kubernetes/kubernetes/issues/69819 .

As @msau42 suggested, we will be able to make this framework called from out-of-tree.
Simplest way would be:
  1. Out-of-tree driver developers define their driver in other places by the same way as test/e2e/storage/drivers/csi.go does, which means implement ```TestDriver``` interface defined in e2e/storage/drivers/base.go and its init function for their drivers.
  2. Also define similar code to test/e2e/storage/csi_volumes.go (after L98 is not necessary and ```csiTestDrivers``` need to be modified to include their driver's init code that is implemented in step 1 above).

We might improve step 2 not to duplicate the code, but its just about 100 steps of code including header and comments, so it should be enough simple.

Also, until moving csi driver test codes other than reference drivers are moved to out-of-tree, we still have a chance to add codes for out-of-tree drivers' e2e tests to in-tree. (As is intended in https://github.com/kubernetes/kubernetes/pull/67088 .)

pohly replied:
This sounds like it could work, thanks for explaining.

mkimuram commented:
@msau42 

> I would leave the volume attachment tests as they are and not move them to the framework. They are not designed to be run against every plugin.

I updated the codes as suggested. PTAL

mkimuram commented:
I seemed to forget to change provisioner name  for gce-pd to "com.google.csi.gcepd". 
Updating the commit.

mkimuram commented:
/test pull-kubernetes-integration

pohly commented on test/e2e/storage/testsuites/subpath.go:
```diff
@@ -315,7 +315,7 @@ func testSubPath(input *subPathTestInput) {
 
 	It("should unmount if pod is force deleted while kubelet is down [Disruptive][Slow]", func() {
-		if input.volType == "hostPath" || input.volType == "hostPathSymlink" {
+		if strings.HasPrefix(input.volType, "hostPath") || strings.HasPrefix(input.volType, "csi-hostpath") {
```
Can this be done without hard-coding the names?

Third-party drivers will be named differently and may also have to skip this test.

mkimuram replied:
@pohly 

Thank you for pointing it out. I will fix it.

@msau42 

Could you explain the reason why this subPath test should be skipped for hostPath, so that I can name the flag in ```TestDriver``` to skip this test properly?
Is it due to non-attachable, local, or lack of reconstruction code?

msau42 replied:
It's a regression due to subpath fixes we made a while back: https://github.com/kubernetes/kubernetes/issues/61446.  It's not intended for some drivers to not support reconstruction and others to support it, so I'm not sure if we really want to make it a parameter.

mkimuram replied:
@msau42 

Thank you for your explanation. I will add a comment there to mark it as TODO with a link to the issue.

mkimuram commented:
Now I understand why csi gce pd fails. It requires more than or equal to 5Gi volume.

Let me change the size of dynamic provision PV to 5Gi for all resource creation.
We might better changing it to tunable for drivers in the future.

```
Oct 25 23:30:48.370: INFO: At 2018-10-25 23:25:29 +0000 UTC - event for pvc-p5g86: {com.google.csi.gcepd csi-gce-controller-0 e8826c17-d8ac-11e8-ae8b-0a580a40002f } ProvisioningFailed: Failed to provision volume with StorageClass "e2e-tests-volumes-mjlkt-com.google.csi.gcepd-sc": rpc error: code = InvalidArgument desc = CreateVolume Request Capacity is invalid: Limit bytes 2147483648 is less than minimum volume size: 5368709120
```

davidz627 commented:
> Now I understand why csi gce pd fails. It requires more than or equal to 5Gi volume.
> 
> Let me change the size of dynamic provision PV to 5Gi for all resource creation.
> We might better changing it to tunable for drivers in the future.
> 
> ```
> Oct 25 23:30:48.370: INFO: At 2018-10-25 23:25:29 +0000 UTC - event for pvc-p5g86: {com.google.csi.gcepd csi-gce-controller-0 e8826c17-d8ac-11e8-ae8b-0a580a40002f } ProvisioningFailed: Failed to provision volume with StorageClass "e2e-tests-volumes-mjlkt-com.google.csi.gcepd-sc": rpc error: code = InvalidArgument desc = CreateVolume Request Capacity is invalid: Limit bytes 2147483648 is less than minimum volume size: 5368709120
> ```

Sorry this caused you issues. The documented minimum size for GCE PD is 5Gi. However in testing I have discovered that it is actually possible to create disks as low as 1Gi. So I have updated the driver to support that in v0.2.0.beta (unreleased)

mkimuram commented:
@davidz627  

Thank you for your comment. I see what is happening there.
It seems that issue with the pv size could be worked around by specifying proper size. So, you don't need to hurry to update the csi GCE PD driver.

However, below two errors still remain.

First issue could be resolve by setting ```IsFsGroupSupported``` to false for csi GCE PD driver, once I confirmed that this driver really doesn't support it. (Is my understanding right?)  
Second issue requires to escape driver name to dnsname (alnum and -). Do you know any good ways to achieve it? 
It would be better to use proper library function, instead of replacing it with regexp, however, I couldn't find a good one. (Or it is just a pod name for the test and test is executed in separated namespace, so we might be able to choose just changing the naming rule for the pod.)

[Errors]
1. csi-gce-pd driver fails in checking fsGroup
Ex) ```[sig-storage] CSI Volumes [Driver: com.google.csi.gcepd] [Testpattern: Dynamic PV (default fs)] volumes should be mountable```

<details>

```
test/e2e/storage/testsuites/volumes.go:151
failed: getting the right privileges in the file 1234
Expected error:
    <*errors.errorString | 0xc001aca330>: {
        s: "Failed to find \"1234\", last result: \"drwxr-xr-x    3 root     root          4096 Oct 26 18:41 /opt/0\n\"",
    }
    Failed to find "1234", last result: "drwxr-xr-x    3 root     root          4096 Oct 26 18:41 /opt/0
    "
not to have occurred
test/e2e/framework/volume_util.go:477
```

</details>

2. Subpath tests name pod by using driver name but pod name doesn't allow '.' because it only allow dnsname.
Ex) ```[sig-storage] CSI Volumes [Driver: com.google.csi.gcepd] [Testpattern: Dynamic PV (default fs)] subPath should support existing directories when readOnly specified in the volumeSource```

<details>

```
test/e2e/storage/testsuites/subpath.go:349
Error creating Pod
Expected error:
    <*errors.StatusError | 0xc001ed2f30>: {
        ErrStatus: {
            TypeMeta: {Kind: "", APIVersion: ""},
            ListMeta: {SelfLink: "", ResourceVersion: "", Continue: ""},
            Status: "Failure",
            Message: "Pod \"pod-subpath-test-com.google.csi.gcepd-dynamicpv-hhgq\" is invalid: [spec.containers[0].name: Invalid value: \"test-container-subpath-com.google.csi.gcepd-dynamicpv-hhgq\": a DNS-1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?'), spec.containers[1].name: Invalid value: \"test-container-volume-com.google.csi.gcepd-dynamicpv-hhgq\": a DNS-1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?'), spec.initContainers[0].name: Invalid value: \"init-volume-com.google.csi.gcepd-dynamicpv-hhgq\": a DNS-1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?')]",
```

</details>

msau42 commented:
1. fsgroup handling is done at the kubernetes csi layer, not the csi driver. so maybe this is a bug with kubernetes csi volume plugin
2. we could regex out the invalid characters, or we can just use a random hash for the pod name?

mkimuram commented:
@msau42 

Thank you for your quick response.

I will mark ```IsFsGroupSupported``` as false for all csi drivers until it is fixed for the first issue,
 and regex out the invalid characters for the second issue.

msau42 commented:
I opened https://github.com/kubernetes/kubernetes/issues/70303 to track csi fsgroup issue

mkimuram commented:
@msau42 @davidz627 

Now all the tests passed. PTAL

davidz627 reviewed (commented):
mostly looks good to me, however still have a few questions/comments

mkimuram commented on test/e2e/storage/testsuites/subpath.go:
```diff
@@ -397,7 +399,9 @@ func TestBasicSubpathFile(f *framework.Framework, contents string, pod *v1.Pod,
 	var (
-		suffix          = strings.ToLower(fmt.Sprintf("%s-%s", volumeType, rand.String(4)))
+		re = regexp.MustCompile("[^[:alnum:]]")
+		// suffix is used in pod name and it needs to be DNS-1123 label, so all strings other than alnum should be replaced with "-"
```
Thank you for your comment.

To avoid below error to happen is the reason. I will try to explain it in the comment.
Also, I reread the below message and found out that we might need to also convert it to lower case. Because ```[[:alnum:]] == [0-9A-Za-z]```, while it requires is ```[a-z0-9]```.

```
"Pod \"pod-subpath-test-com.google.csi.gcepd-dynamicpv-hhgq\" is invalid: [spec.containers[0].name: Invalid value: \"test-container-subpath-com.google.csi.gcepd-dynamicpv-hhgq\": a DNS-1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?'), spec.containers[1].name: Invalid value: \"test-container-volume-com.google.csi.gcepd-dynamicpv-hhgq\": a DNS-1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?'), spec.initContainers[0].name: Invalid value: \"init-volume-com.google.csi.gcepd-dynamicpv-hhgq\": a DNS-1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?')]"
```

mkimuram replied:
Oh, ```ToLower()``` has already been done there.

mkimuram commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -40,96 +39,81 @@ import (
-				driver.createCSIDriver()
+				// setupDriver
+				drivers.SetCommonDriverParameters(driver, f, config)
+				driver.CreateDriver()
```
@msau42 

Thank you for your answer. 
Is my understanding correct that it is also true for all the csi drivers including csi-hostpath?

Then, I will add ```[Serial]``` to test execution logic in csi_volumes.go instead of each ```TestDriver```'s ```FeatureTag```.

msau42 replied:
@pohly is in the process of making hostpath parallelizable, so only gce pd needs to be serialized.

mkimuram commented:
/retest

mkimuram commented:
It seems that #69868 is about to merge, I'll wait for it and rebase to the latest.

davidz627 commented:
@mkimuram were you able to test the GCE PD CSI Driver tests with this PR?

mkimuram commented:
Just for your information https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/68025/pull-kubernetes-e2e-gce/57308/ is the test log for commit eef53a2, which made all gce pd test pass.

pohly commented:
Masaki Kimura <notifications@github.com> writes:
> It seems that #69868 is about to merge, I'll wait for it and rebase to
> the latest.

Thanks. It has been merged now, please rebase.

davidz627 commented:
Will do one last pass later on in the day, for now
/hold
so I can test the GCE PD CSI tests

mkimuram commented:
@davidz627 

Thank you for your help in testing in gce environment.

gnufied commented:
@davidz627 gce-pd tests are still very flaky because of https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/issues/144 not sure if you are getting the same. But since these tests won't run in all environments may be we are okay?

gnufied commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -0,0 +1,284 @@
+
+// InitGcePDExternalCSIDriver returns gcePDExternalCSIDriver that implements TestDriver interface
+func InitGcePDExternalCSIDriver() TestDriver {
+	return &gcePDCSIDriver{
```
If I am reading this correctly, this defeats the point of external CSI driver test, because we are not returning a instance of `gcePDCSIDriver` which has concrete implementation of `CreateDriver` call. Is that what we want?

mkimuram commented:
@gnufied 

Thank you for your comment, let me update the code to re-enable fsGroup test.

mkimuram commented:
Updated.

davidz627 commented:
@mkimuram I've noticed when testing that for all the tests that are `skipped` the driver deploy and teardown is still running. This causes each skipped test to take quite a while to skip. If we determine that the test does not need to run can we also skip the step of deploying the driver

mkimuram commented:
@davidz627 

You're right that current code will make all ```driver.CreateDriver()``` to be called for any tests including those will be skipped. It is due to the fact that driver creation code exists in outer loop and I couldn't find a good way to skip the tests there, because there seems no way to know which tests will actually be executed or skipped in the outer loop. Any ideas to improve it?

mkimuram commented:
@davidz627

Let me add some comments on it.

There are many tests skipped currently, because only dynamic provisioner cases are executed, due to lack of implementation for inline volume and pre-previsioned volume in CSI TestDrivers. However, once these codes are implemented, most of the tests will be executed. So, I would expect that the overhead of the skips won't be a big matter. 
In addition, I understand that @msau42 is thinking about making driver creation code run once in entire execution, when it is moved to each csi driver's repository as a library, if it is achieved this shouldn't be an issue. (Also, doing it while it is still in k/k repository could be an option.)

msau42 commented:
I don't think we need to test inline cases for CSI drivers.  Can we at least remove that test pattern for the CSI tests?

mkimuram commented:
@msau42 

Rebased and removed the inline volume cases from CSI driver tests.

msau42 commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -58,6 +59,20 @@ var csiTestSuites = []func() testsuites.TestSuite{
+
+	for _, pattern := range patterns {
+		// Skip inline volume tests
+		if pattern.VolType == testpatterns.InlineVolume {
```
I just discussed with @davidz627, let's also skip the pre-provisioned case too because there's some 50-ish test cases.  sig-testing is encouraging moving provider-specific test cases out of tree, so if we decide to add test cases for pre-provisioned volumes in the future, we should do it out-of-tree.

mkimuram replied:
@msau42 

I agree with it. I will update the codes as so.

msau42 commented:
/approve

Thanks so much for working on this! This really helps increase our CSI testing coverage (from 1 test case to ~50!)

mkimuram commented:
@msau42 @davidz627 

Changed to also skip pre-provisioned PV tests for CSI and squashed the commits. PTAL

pohly commented:
Masaki Kimura <notifications@github.com> writes:
> @davidz627 
>
> You're right that current code will make all
> ```driver.CreateDriver()``` to be called for any tests including those
> will be skipped. It is due to the fact that driver creation code
> exists in outer loop and I couldn't find a good way to skip the tests
> there, because there seems no way to know which tests will actually be
> executed or skipped in the outer loop. Any ideas to improve it?

You could just define init and cleanup functions in the outer BeforeEach
and then pass those to `testsuites.RunTestSuite`. Then if (and only if)
a test actually runs, it has to call the init function and later the
cleanup.

mkimuram commented:
@pohly 

Thank you for sharing your idea.
I confirmed that below patch will make driver setup skipped if the test is skipped.

However, I guess that it is going to a bit different direction to @msau42 's plan to make driver setup run once in entire test runs. Although, it looks enough small change to be reverted easily in the future, so it might be useful for current framework, until we achieve the plan.

@msau42 

How do you think of this patch? 

<details>

```
diff --git a/test/e2e/storage/csi_volumes.go b/test/e2e/storage/csi_volumes.go
index f86d411257..d71eff1b22 100644
--- a/test/e2e/storage/csi_volumes.go
+++ b/test/e2e/storage/csi_volumes.go
@@ -134,14 +134,7 @@ var _ = utils.SIGDescribe("CSI Volumes", func() {
 			driver := curDriver
 
 			BeforeEach(func() {
-				// setupDriver
 				drivers.SetCommonDriverParameters(driver, f, config)
-				driver.CreateDriver()
-			})
-
-			AfterEach(func() {
-				// Cleanup driver
-				driver.CleanupDriver()
 			})
 
 			testsuites.RunTestSuite(f, config, driver, csiTestSuites, csiTunePattern)
diff --git a/test/e2e/storage/in_tree_volumes.go b/test/e2e/storage/in_tree_volumes.go
index 6c6ea61368..975016f9fa 100644
--- a/test/e2e/storage/in_tree_volumes.go
+++ b/test/e2e/storage/in_tree_volumes.go
@@ -81,14 +81,7 @@ var _ = utils.SIGDescribe("In-tree Volumes", func() {
 			driver := curDriver
 
 			BeforeEach(func() {
-				// setupDriver
 				drivers.SetCommonDriverParameters(driver, f, config)
-				driver.CreateDriver()
-			})
-
-			AfterEach(func() {
-				// Cleanup driver
-				driver.CleanupDriver()
 			})
 
 			testsuites.RunTestSuite(f, config, driver, testSuites, intreeTunePattern)
diff --git a/test/e2e/storage/testsuites/base.go b/test/e2e/storage/testsuites/base.go
index 8d4f084989..a1f0e2e617 100644
--- a/test/e2e/storage/testsuites/base.go
+++ b/test/e2e/storage/testsuites/base.go
@@ -135,6 +135,9 @@ var _ TestResource = &genericVolumeTestResource{}
 
 // setupResource sets up genericVolumeTestResource
 func (r *genericVolumeTestResource) setupResource(driver drivers.TestDriver, pattern testpatterns.TestPattern) {
+	// SetupDriver
+	driver.CreateDriver()
+
 	r.driver = driver
 	dInfo := driver.GetDriverInfo()
 	f := dInfo.Framework
@@ -223,6 +226,9 @@ func (r *genericVolumeTestResource) cleanupResource(driver drivers.TestDriver, p
 
 	// Cleanup volume for pre-provisioned volume tests
 	drivers.DeleteVolume(driver, volType, r.driverTestResource)
+
+	// Cleanup driver
+	driver.CleanupDriver()
 }
 
 func createVolumeSourceWithPVCPV(
diff --git a/test/e2e/storage/testsuites/provisioning.go b/test/e2e/storage/testsuites/provisioning.go
index f897f54beb..764926ab56 100644
--- a/test/e2e/storage/testsuites/provisioning.go
+++ b/test/e2e/storage/testsuites/provisioning.go
@@ -142,6 +142,9 @@ type provisioningTestResource struct {
 var _ TestResource = &provisioningTestResource{}
 
 func (p *provisioningTestResource) setupResource(driver drivers.TestDriver, pattern testpatterns.TestPattern) {
+	// SetupDriver
+	driver.CreateDriver()
+
 	// Setup provisioningTest resource
 	switch pattern.VolType {
 	case testpatterns.DynamicPV:
@@ -162,6 +165,8 @@ func (p *provisioningTestResource) setupResource(driver drivers.TestDriver, patt
 }
 
 func (p *provisioningTestResource) cleanupResource(driver drivers.TestDriver, pattern testpatterns.TestPattern) {
+	// Cleanup driver
+	driver.CleanupDriver()
 }
 
 type provisioningTestInput struct {
diff --git a/test/e2e/storage/testsuites/volumemode.go b/test/e2e/storage/testsuites/volumemode.go
index 360097b66f..1892bff14c 100644
--- a/test/e2e/storage/testsuites/volumemode.go
+++ b/test/e2e/storage/testsuites/volumemode.go
@@ -154,6 +154,9 @@ type volumeModeTestResource struct {
 var _ TestResource = &volumeModeTestResource{}
 
 func (s *volumeModeTestResource) setupResource(driver drivers.TestDriver, pattern testpatterns.TestPattern) {
+	// SetupDriver
+	driver.CreateDriver()
+
 	s.driver = driver
 	dInfo := driver.GetDriverInfo()
 	f := dInfo.Framework
@@ -226,6 +229,9 @@ func (s *volumeModeTestResource) cleanupResource(driver drivers.TestDriver, patt
 
 	// Cleanup volume for pre-provisioned volume tests
 	drivers.DeleteVolume(driver, volType, s.driverTestResource)
+
+	// Cleanup driver
+	driver.CleanupDriver()
 }
 
 type volumeModeTestInput struct {
```

msau42 commented:
I prefer not to move driver setup into test cases for that reason.  Once CSI is stable and migration is enabled, providers have package in their drivers as part of cluster deployment, so all this logic for starting up drivers in e2e tests will be unnecessary.

mkimuram commented:
@pohly 

> It would be good if testsuites.RunTestSuite supported both modes

This may improve the execution time of e2e tests, however we should do it in a separate PR, because
over head by skipping tests has already been worked around by not executing inline volume and pre-provisioned PV, so it is not a blocker for this PR now.

davidz627 commented:
running the GCE PD CSI tests with the most up to date version of this PR right now, will update ASAP

pohly commented:
Masaki Kimura <notifications@github.com> writes:
>> It would be good if testsuites.RunTestSuite supported both modes
>
> This may improve the execution time of e2e tests, however we should do
> it in a separate PR, because over head by skipping tests has already
> been worked around by not executing inline volume and pre-provisioned
> PV, so it is not a blocker for this PR now.

Yes, that's fine with me.

davidz627 commented:
Ran all the GCE tests, log here: https://drive.google.com/open?id=19_fPVcKYtq_vqkhXOk7ZntMq8ueEJXzY

had a couple test failures:
```
Summarizing 3 Failures:

[Fail] [sig-storage] CSI Volumes [Driver: com.google.csi.gcepd][Serial] [Testpattern: Dynamic PV (block volmode)] volumeMode[Feature:BlockVolume] [It] should fail in binding dynamic provisioned PV to PVC 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/volumemode.go:331

[Fail] [sig-storage] CSI Volumes [Driver: com.google.csi.gcepd][Serial] [Testpattern: Dynamic PV (default fs)] subPath [It] should fail for new directories when readOnly specified in the volumeSource [Slow] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/subpath.go:615

[Fail] [sig-storage] CSI Volumes [Driver: com.google.csi.gcepd][Serial] [Testpattern: Dynamic PV (default fs)] subPath [It] should support existing directories when readOnly specified in the volumeSource 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2369
```

Looks like this boils down to:
1. No block volume support <- not supported yet, tests should be removed/skipped
2. Read only in the volume source issues <- this is a real problem, needs to be fixed

Do we will want to merge knowing these are real failures, or should we resolve them first and hold this PR?
I am amenable to resolving the block volume issue first and dealing with the `readOnly` issue after merging this
@msau42 @mkimuram

msau42 commented:
I remember discussing with @saad-ali there is a lot of confusion and maybe some issues around the read only flag.  Let me  see if we already have bugs filed

msau42 commented:
https://github.com/kubernetes/kubernetes/issues/70503 is mainly talking about confusion.  But it seems like there could be a real bug here, so maybe we need a separate issue for that.  It's odd that this is passing for hostpath and not for gce pd.

mkimuram commented:
@davidz627 

Thank you for your test and comment.

> I think we should skip the block volume test, thenlet this PR merge. 

What issue did you see in block volume test?

I expect that it is skipped in environments where ```BlockVolume``` is not enabled, because ```[Feature:BlockVolume]``` featureTag is set to the test suite.
Also, for as long as ``` isBlockSupported``` is marked as ```false``` for a driver, the driver should be regarded as block unsupported, so it should work as expected. (As it works well for in-tree hostpath driver or nfs driver.)

Because of the second reason, I also expect that even when ```BlockVolume``` become default to enabled and ```CSIBlockVolume``` remains default to disabled, it would work well if ``` isBlockSupported``` is marked as ```false``` for all csi drivers .

If above is not enough, I will consider hardcoding csi gce to skip by implementing ```skipUnsupportedTest``` for ```volumeModeTestSuite```.

mkimuram commented:
As for the block volume test, I found below log and it indicates that the test expects error because the driver doesn't support block volume, but no error happened. So, I suspect that it was executed in an environment where ```BlockVolume``` is not enabled.

```
  [Driver: com.google.csi.gcepd][Serial]
  [90m/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/csi_volumes.go:133[0m
    [Testpattern: Dynamic PV (block volmode)] volumeMode[Feature:BlockVolume]
    [90m/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/volumemode.go:110[0m
      [91m[1mshould fail in binding dynamic provisioned PV to PVC [It][0m
      [90m/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/volumemode.go:316[0m

      [91mExpected an error to have occurred.  Got:
          <nil>: nil[0m

      /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/volumemode.go:331
```

[It](https://github.com/kubernetes/kubernetes/blob/5047e6ec263ca37282ca39478583d96eb8883ca8/test/e2e/storage/testsuites/volumemode.go#L331) actually expects dynamic-provisioned PV for block unsupported drivers not to be bound to PVC if volume mode doesn't match, but in non block supported environment, block volume mode is not handled, the check won't work well.

davidz627 commented:
Thanks for the investigation and the great PR @mkimuram !!
We will need to fix the `readOnly` issue before this release since these tests will all become release blocking (https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/issues/147).
We also need to remove `[Feature:BlockVolume]` tests from the GCE PD CSI Driver repo (https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/issues/146)
/lgtm
/hold cancel

msau42 commented:
Can you also open an issue for in-tree tests?

msau42 commented:
/hold
actually on second thought, I think we should look into fixing it before we merge this.  It will cause less headache for the release team.

AishSundar commented:
/milestone v1.13
/cc @nikopen

msau42 commented:
Can you try running `subPath [It] should fail for new directories when readOnly specified in the volumeSource [Slow] ` against the hostpath driver?

mkimuram commented:
> Can you try running subPath [It] should fail for new directories when readOnly specified in the volumeSource [Slow] against the hostpath csi driver?

I confirmed that this test also fails for csi hostpath.

In in-tree, this test for hostpath is skipped by not returning readOnly in-line volume source. ([skip](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/subpath.go#L364), [volumeSource](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/in_tree.go#L703))This could be achieved because in-tree hostpath doesn't support dynamic provisioning, but we currently don't have a way to skip this test for [dynamic provisioning](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/subpath.go#L167) (and [pre-provisioned PV](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/subpath.go#L160)), so we need to implement some code, if we need to skip it for csi-hostpath.

msau42 commented:
I think hostpath CSI should be able to support this. @davidz627 debugged a little yesterday, and he found out that the way the Kubernetes mount library does the read-only bind mount does not propogate to the host.  We're investigating if there's other ways to do a read only bind mount inside a container

mkimuram commented:
Cool! I will wait for it.

msau42 commented:
In the meantime, can you also run the other slow, serial hostpath csi tests to see if there could be other problems?

mkimuram commented:
OK. I will run them and share the results.

mkimuram commented:
By the way, this could be a separated topic to this PR. I found that we would be still nice to have improvement similar to https://github.com/kubernetes/kubernetes/pull/68689 . That subpath test still wait the pod to be [running](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/subpath.go#L589), while the pod remains ```CreateContainerConfigError``` even after the expected error event happens. Once it is achieved, we would be able to untag slow from this test.

msau42 commented:
Does https://github.com/kubernetes/kubernetes/pull/69300 help with shortening the test time?

mkimuram commented:
> Does #69300 help with shortening the test time?

No. I executed this test with in-tree iscsi several times, but it seems to wait 5 minutes. And while the test is running, I confirmed that the pod's status remains ```CreateContainerConfigError```.
So, I think that we should wait for the pod to be ```CreateContainerConfigError``` not ```Running``` or timeout.

mkimuram commented:
Or just deleting ```WaitForPodRunningInNamespace``` might be fine.

@wongma7 

Did you mean above by "I think waiting "WaitForPodRunning" to timeout is still redundant" in #68689 ?

mkimuram commented:
@msau42 

For 11 slow tests and 0 serial tests for csi-hostpath, 8 test passed, 1 test skipped, and 2 tests failed. 
[csi-hostpath-slow-testresult.txt](https://github.com/kubernetes/kubernetes/files/2543479/csi-hostpath-slow-testresult.txt)

As for failed tests, one is the test already mentioned above. And the other is disruptive test that require external IP to make node down, which can't be executed in my environment.

So, the second failed test should be tested by someone else. Then, we will be sure that all tests are good.

<details>

[Slow tests]
```
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] subPath should unmount if pod is force deleted while kubelet is down [Disruptive][Slow]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] volumeIO should write files of various sizes, verify size, validate content [Slow]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] subPath should fail if subpath directory is outside the volume [Slow]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] subPath should fail for new directories when readOnly specified in the volumeSource [Slow]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] subPath should support creating multiple subpath from same volumes [Slow]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] subPath should fail if subpath with backstepping is outside the volume [Slow]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] subPath should fail if non-existent subpath is outside the volume [Slow]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] subPath should unmount if pod is gracefully deleted while kubelet is down [Disruptive][Slow]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] subPath should support restarting containers using file as subpath [Slow]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] subPath should support restarting containers using directory as subpath [Slow]
[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] subPath should fail if subpath file is outside the volume [Slow]
```

[Serial tests]
```
None
```

[Failed tests]
```
[Fail] [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] subPath [It] should fail for new directories when readOnly specified in the volumeSource [Slow] 
/root/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/subpath.go:615
--
[Fail] [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] subPath [It] should unmount if pod is gracefully deleted while kubelet is down [Disruptive][Slow] 
/root/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/utils/utils.go:206
```

mkimuram commented:
@msau42 @davidz627 

Rebased the commit. PTAL
Only the change applied was deletion of "controller_service.yaml" in test/e2e/storage/drivers/csi.go, as it was changed in test/e2e/storage/csi_volumes.go in commit 45ee18a.

mkimuram commented:
/retest

davidz627 commented:
/lgtm

msau42 commented:
/hold cancel

mkimuram commented:
/retest

AishSundar commented:
@jberkus @mariantalla @mortent 

FYI to watch for any possible CI instability post this merge.

msau42 commented:
There is one known test failure that will be addressed by: #70662

AishSundar commented:
/priority important-soon

AishSundar commented:
@msau thanks. Looks like https://github.com/kubernetes/kubernetes/pull/70662 is still pending approval. are you intentionally synchronizing the merges?

msau42 commented:
@AishSundar we can hold the merge on this until #70662 merges if you think that will save the release team some effort in triaging.

AishSundar commented:
yes plz, if #70662  addresses the instability / test failure that will surface once this PR is merged and if #70662 can be merged preemptively before this,  then lets order them that way. Thanks.

msau42 commented:
/hold

msau42 commented:
/hold cancel
/retest

mkimuram commented:
/retest

mkimuram committed:
Make csi drivers and in-tree drivers share e2e tests

mkimuram commented:
Rebased to avoid conflict to a679486.
The difference is just removing ```createCSICRDs``` as it is done in the above commit.

mkimuram commented:
SkipAttach test doesn't seem to work properly in my environment. I'm still looking into it.

mkimuram commented:
@saad-ali 

It seems that csi driver crd resource is not created, after I merged this PR with a679486,
as a result, creating [csi driver cr](https://github.com/kubernetes/kubernetes/pull/68025/commits/b89b367247039f4a46a1ca2f0e4e906d723d79d2#diff-0d9ecaa3e6a0297186ad33f57aad472eR195) fails in skipAttach test.
Could you help me debugging this?

I couldn't figure out when and how ```TestNodeAuthorizer``` is called, so I found difficulty in finding root cause.

msau42 commented:
Opened up https://github.com/kubernetes/kubernetes/issues/70760 for alpha test issues

pohly commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -0,0 +1,281 @@
+	parameters := map[string]string{"type": "pd-standard"}
+
+	return getStorageClass(provisioner, parameters, nil, ns, suffix)
+}
```
@mkimuram Should this gcePDCSIDriver.GetDynamicProvisionStorageClass really ignore the fsType string? The driver declares that it supports multiple different filesystems.

Does this lead to multiple different test configurations (one for each filesystem) where each individual test run then uses the same default fs?

mkimuram replied:
@pohly 

Thank you for pointing it out. Current ```gcePDCSIDriver.GetDynamicProvisionStorageClass``` code lacks setting ```fsType``` as [it was done for in-tree one](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/in_tree.go#L1176). Also, I came to think that volumes test should test if the volume is mounted with the proper fsType, if fsType is specified. (Currently, it [tests multiple fsType](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/volumes.go#L44) but doesn't seem to [test fsType](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/volume_util.go#L404).)

I will make a PR to fix it.

mkimuram replied:
Opened an issue and a PR below:

Issue: https://github.com/kubernetes/kubernetes/issues/70822
PR: https://github.com/kubernetes/kubernetes/pull/70823

pohly commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -0,0 +1,281 @@
+			),
+			IsPersistent:       true,
+			IsFsGroupSupported: false,
+			IsBlockSupported:   false,
```
@mkimuram it's not obvious what "IsPersistent" means, and the description ("Flag to represent whether it provides persistency") doesn't explain it either (it just paraphrases the variable name, which usually is a comment that might as well be left out because it doesn't add additional information).

Does it mean "persistent across node restarts"? The hostpath driver isn't that, so I'm wondering why it is set to true here.

I also don't know what "fsGroup" is.

"Block support" is a bit easier to guess, but an explicit description would be better.

mkimuram replied:
@pohly 

> it's not obvious what "IsPersistent" means

Actually, I meant "persistent across pod restarts".  It is defined to handle [emptyDir driver](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/in_tree.go#L899
) properly in [volumes test](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/volumes.go#L72). I agree that we should improve the comment and we might need something like ```IsPersistentAcrossNode``` flag in the future.

> I also don't know what "fsGroup" is

Please see below.
https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#set-the-security-context-for-a-pod

Most of these flags are defined in the process of migrating existing in-tree driver's test to this framework. So, we will need more flags in the future and improve these flags and their comments.

msau42 replied:
I'm changing the way these capabilities are defined and hopefully making the comments more clear here: https://github.com/kubernetes/kubernetes/pull/70756

mkimuram replied:
@msau42 

Thank you for start working on improving capability definition.
The framework is ready for both in-tree and csi now, so it would be good timing to do it, before adding more testsuites and drivers.

pohly commented:
Thanks for getting this merged. I didn't have time earlier, but now (better late than never) I am trying to use the testsuite package out-of-tree, which led to some questions.

pohly commented on test/e2e/storage/testsuites/base.go:
```diff
@@ -66,12 +66,12 @@ func getTestNameStr(suite TestSuite, pattern testpatterns.TestPattern) string {
 
 // RunTestSuite runs all testpatterns of all testSuites for a driver
-func RunTestSuite(f *framework.Framework, config framework.VolumeTestConfig, driver drivers.TestDriver, tsInits []func() TestSuite) {
+func RunTestSuite(f *framework.Framework, config framework.VolumeTestConfig, driver drivers.TestDriver, tsInits []func() TestSuite, tunePatternFunc func([]testpatterns.TestPattern) []testpatterns.TestPattern) {
```
Is framework.VolumeTestConfig a good choice here? It has lots of options that don't seem to be used anymore, or at least don't have an effect. For example, provisioning.go only uses ClientNodeName but ignores NodeSelector.

Do you think it might be clear to have a testsuite.Config which has just the options that really matter for the testsuite package?

mkimuram replied:
Good points. 

This framework has created in a way that existing tests can be called in a common way for any drivers. (These tests were originally tightly bound to particular driver's particular source, like volume source, pv source, or storage class. And this PR was aimed to separate them to drivers and testsuites, and make testsuites called in a common loop.) 

In the process, I tried to keep the existing tests as much as they were so that I won't break how tests were working. In other word, how the test is called was changed, but how each test works haven't been improved, yet. 

Now, above aim was achieved, so we are ready to consider improving how each test works. Defining a new testsuite.Config that can be commonly used in all testsuites will be one of the remaining items on how each test works. And it will be helpful in adding more testsuites easily without bugs in the future.

pohly commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -151,7 +151,7 @@ func (p *provisioningTestResource) setupResource(driver drivers.TestDriver, patt
 			}
 			p.driver = driver
-			p.claimSize = "2Gi"
+			p.claimSize = "5Gi"
```
@mkimuram This hard-coded size doesn't work for me: my test setup provisions in RAM and can't support such large volumes. This needs to be configurable.

I can work on a PR for this, but let's clarify first what the role of `framework.VolumeTestConfig` is in `testsuites`.

Speaking of volume size, the subpath tests fail for the same reason. But I haven't figured out yet where that PVC gets created. Can you point me in the right direction?

mkimuram replied:
@pohly 

This value was changed to 5Gi due to the restriction of current version of csi gce driver.
Once @davidz627 push the latest version of csi gce driver, then we will be able to change it to smaller value.

Also, we will be able to tune it by defining something like minimum volume size in each driver and set the size here.

> But I haven't figured out yet where that PVC gets created. Can you point me in the right direction?

It depends on the testsuites but it should be created either in
1. Each tests ([Inside ```It()```](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/provisioning.go#L215)) (ex. provisioning, volumeMode testsuites)
2. [setupResource](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/base.go#L137) (ex. subPath, volumes, volumeIO testsuites)

It is different because 
 - if the scope of the test includes the creation of a resource, it should be inside ```It()```,
 - if the scope of the test doesn't include the creation of a resource, it should be in ```setupResource``` to avoid duplicated logic in all tests. (Most of tests will be able to use genericVolumeTestResource as is.)

msau42 replied:
I agree on making PVC size configurable.  For example, Google Filestore volumes have a min size of 1TB

pohly commented:
When trying to use the `testsuite` package out-of-tree for my own CSI driver, it ended up pulling in the `drivers` package (because of the interface) and all in-tree driver definitions (because they are in the same package). That in turn pulled in tests from storage/vsphere which my test suite is never going to run. The driver interface should be in a separate package to avoid that.

I propose that we structure it like this:
- `storage/drivers` - the in-tree driver definitions for use in Kubernetes E2E, as they are now
- `storage/testsuites/driver` - just the interface definition (new package)
- `storage/testsuites` - the current tests

The rationale for that is that a) `storage/testsuites`as the main (only?!) user of the interface should own it b) having it as a separate package might simplify driver definitions which don't need the full `testsuites` package

@msau42, @mkimuram if you agree then I can prepare a PR.

mkimuram commented:
@pohly 

If you are working on utilizing this framework on a certain csi driver's, I think that it would be better to create a separate issue/PR for the driver and share the code and the results. Then, I think that we can discuss the specific issue on it there. Also, if there will be some needs in changing the code in k8s side, then we will need to create a separate issue/PR for k8s.

pohly commented:
Masaki Kimura <notifications@github.com> writes:
> If you are working on utilizing this framework on a certain csi
> driver's, I think that it would be better to create a separate
> issue/PR for the driver and share the code and the results.

I beg to differ here. The Kubernetes CI is not the right place to test
everything related to Kubernetes. CSI drivers are by design developed
outside of Kubernetes, therefore developers of such a driver also need a
way to test their driver without modifying Kubernetes source code.

Making the in-tree tests configurable at runtime is one possibility, but
then running the tests is still cumbersome (must build e2e.test in
Kubernetes or use some pre-compiled version). I'd rather vendor in the
test suite and the shared tests and integrate them with some other,
driver specific tests into my own test suite binary.
